ComponentKit/Utilities/CKComponentTrigger.h

```diff
@@ -0,0 +1,169 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <Foundation/Foundation.h>
+
+#import <memory>
+
+#import <ComponentKit/CKComponentScope.h>
+#import <ComponentKit/CKComponentAction.h>
+
+/**
+ A component trigger is the opposite of an action. It is used by a parent component to trigger some side effect in a
+ child component.
+ 
+ Similar to an action, it is created in a parent scope, and is threaded down to a child component. Unlike an action
+ though, triggers activate **downward** from a parent component onto a child. Actions are triggered by a child, and
+ cause side-effects in parent components.
+ 
+ Triggers are an alternative to manually exposing methods on parent components that invoke on child components. Instead
+ triggers can be passed through many layers of components.
+ 
+ Triggers use the same pattern as component actions to allow parameters to be passed to targets. The first argument to
+ a trigger selector will always be the sender of the trigger invocation, the next arguments types are determined by
+ the variadic template types. This allows passing of data down to the trigger invocation method.
+ 
+ Triggers passed into a component **must** be resolved once, and only once. Resolution assigns a target and selector
+ with the trigger for invocation. Here's an example:
+ 
+     + (instancetype)newWithTriggerHandle:(CKTypedComponentTriggerHandle<BOOL>)someTriggerHandle
+     {
+       CKComponentScope scope(self);
+       if (someTriggerHandle) {
+         someTriggerHandle.resolve(scope, @selector(someMethodWithSender:booleanParameter:));
+       }
+       return [super new...];
+     }
+ 
+ Component triggers may be resolved in three different ways:
+ 
+   1. (Preferred) Scope resolution. Allows a CKComponent to register itself or its controller with the trigger.
+      If the selector is not implemented on the component itself, it will be invoked on the controller associated with
+      that component. At runtime in DEBUG mode the trigger will validate that your component or controller implements
+      the selector, and that the parameter types match.
+ 
+       CKComponentScope scope(self);
+       someTriggerHandle.resolve(scope, @selector(someMethodWithSender:booleanParameter:));
+
+   2. (Discouraged) Target resolution. Allows an arbitrary object to receive a trigger invocation. This object is weakly
+      held by the trigger.
+ 
+       Foo *foo = ...;
+       someTriggerHandle.resolve(foo, @selector(someMethodWithSender:booleanParameter:));
+ 
+   3. No-op resolution. Since triggers *must* be resolved, this allows a component to resolve the trigger with no
+      target. This means when the trigger is invoked, nothing will be called.
```

Comment:
Can you provide an example when this form of resolution would be necessary? I feel like a trigger must always be properly resolved, otherwise the contract with the parent invoking the trigger is broken.